### DISCUSSION TITLE
ch1-basic/ch1-03-array-string-and-slice

### DISCUSSION THREAD

**Discussion by giscus** [chai2010/advanced-go-programming-book#622] [General]
# ch1-basic/ch1-03-array-string-and-slice

Go语言高级编程

https://chai2010.cn/advanced-go-programming-book/ch1-basic/ch1-03-array-string-and-slice.html

**Comment by PaiGack** [comment]
成功数 = copy(目标数据, 原数据)

**Comment by wan2435203490** [comment]
SortFloat64FastV这种排序有问题，float64类型指数的起始数为011 1111 1111, -1.0内存里存的是1 011 1111 1111 0....,-2.0存的是1 100 0000 0000 0....,这就导致了排序后-2.0>-1.0

**Reply by gocpplua** [reply]
你的机器32位 还是 64位

**Reply by TwinklerG** [reply]
64位机器的int也是32位的

**Comment by Cyberpunk314** [comment]
func main() {
	a := []float64{1.1, -1.1, 0, 1, 2, 3, 4, -10}
	SortFloat64FastV2(a)
	fmt.Println(a)
}

func SortFloat64FastV2(a []float64) {
    // 通过 reflect.SliceHeader 更新切片头部信息实现转换
    var c []int
    aHdr := (*reflect.SliceHeader)(unsafe.Pointer(&a))
    cHdr := (*reflect.SliceHeader)(unsafe.Pointer(&c))
    *cHdr = *aHdr

    // 以 int 方式给 float64 排序
    sort.Ints(c)
}

output: [-1.1 -10 0 1 1.1 2 3 4]

确实有问题

**Comment by miyu-miyu** [comment]
int 不是 32bit吗？为什么可以吧float64转换成int？

**Reply by miyu-miyu** [reply]
没注意到注释说用的是64位机器，当我没说过

**Reply by miziaiba** [reply]
64位机器，int也是32位啊，long才是64位

**Reply by mck753** [reply]
@miziaiba 64位机器，int不是64位吗

**Comment by 8vd8** [comment]
写的还是很好的

**Comment by arvin-yd** [comment]
用 for range 方式迭代的性能可能会更好一些

**Reply by arvin-yd** [reply]
针对这个问题，特地做了一下测试，fori的性能更好一些
